src/librustc/traits/select.rs

```diff
@@ -677,14 +678,25 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
     fn evaluate_predicate_recursively<'o>(
         &mut self,
         previous_stack: TraitObligationStackList<'o, 'tcx>,
-        obligation: &PredicateObligation<'tcx>,
+        obligation: PredicateObligation<'tcx>,
     ) -> Result<EvaluationResult, OverflowError> {
-        debug!("evaluate_predicate_recursively({:?})", obligation);
+        debug!("evaluate_predicate_recursively(previous_stack={:?}, obligation={:?})",
+            previous_stack.head(), obligation);
+
+        // Previous_stack stores a TraitObligatiom, while 'obligation' is
+        // a PredicateObligation. These are distinct types, so we can't
+        // use any Option combinator method that would force them to be
+        // the same
+        match previous_stack.head() {
+            Some(h) => self.check_recursion_limit(&obligation, h.obligation)?,
+            None => self.check_recursion_limit(&obligation, &obligation)?
+        }
 
         match obligation.predicate {
             ty::Predicate::Trait(ref t) => {
                 debug_assert!(!t.has_escaping_bound_vars());
-                let obligation = obligation.with(t.clone());
+                let mut obligation = obligation.with(t.clone());
+                obligation.recursion_depth += 1;
```

Comment:
I don't think we need to increase the recursion depth here -- this is just translating the predicate from one form to another.

Reply:
I believe it is necessary here. Consider the following call stack:
1. `evaluate_predicate_recursively`
2. `evaluate_trait_predicate_recursively`
3. `evaluate_stack`
4. `evaluate_candidate`
5. `evaluate_predicates_recursively`
6. `evaluate_predicate_recursively`

Here, we've recursed back to `evaluate_predicate_recursively` without ever incrementing the recursion depth. To ensure that repeatedly recursing through the `ty::Predicate::Trait` match arm doesn't lead to a stack overflow, we need to increment the depth here.

Reply:
I think that there should be an increase in between steps 4 and 5. In particular, the "nested obligations" returned by `evaluate_candidate` ought to be at a recursion level one higher than the candidate that spawned them.